### ISSUE TITLE
Examples errors

### ISSUE THREAD

**Issue by olegabr** [qcubed/framework#869]
- [X] `assets/php/examples/code_generator/intro.php` The picture of the db structure is outdated
 - [X] `profile.php` in the value of field `Profile Generated From` HtmlEntities function scramble text making it human unreadable.
 - [X] `assets/php/examples/more_codegen/custom.php` `Undefined GET property or variable in 'Project' class: fltBudget` It is becase Budget is generated with type string, not float. In the `assets/php/examples/mysql_innodb.sql` file it has type `DECIMAL`: `budget DECIMAL`
 - [X] `assets/php/examples/more_codegen/optimistic_locking.php` `Attempting to save a Two Instances of the Same Object will throw an Optimistic Locking Exception` if this button is pressed, error page is shown. Alert box would be helpfull here:
```
Exception in PHP Script /example.com/vendor/qcubed/framework/assets/php/examples/more_codegen/optimistic_locking.php
Optimistic Locking constraint when trying to update PersonWithLock object.  To update anyway, call ->Save() with $blnForceUpdate set to true

Exception Type: QOptimisticLockingException
```

 - [X] `assets/php/examples/qcubed_query/qcache.php` Press the `Clear the cache` button. Message is shown: `Clearing the query cache for the Person table was NOT successful - check your cache / namespace paths. Reload the page to see the effect - query will not be executed against the database.`
 - [X] On any example page in the generated code:
```html
		<link rel="stylesheet" type="text/css" href="&sol;example&period;com&sol;vendor&sol;qcubed&sol;framework&sol;assets&sol;css&sol;styles&period;css"></link>
		<link rel="stylesheet" type="text/css" href="&sol;example&period;com&sol;vendor&sol;qcubed&sol;framework&sol;assets&sol;php&sol;examples&sol;includes&sol;examples&period;css"></link>

		<div id="tagline"><a href="http://qcubed.github.com/" title="QCubed Homepage"><img id="logo" src="&sol;example&period;com&sol;vendor&sol;qcubed&sol;framework&sol;assets&sol;images&sol;qcubed&lowbar;logo&lowbar;footer&period;png" alt="QCubed Framework" /> <span class="version">3&period;0&period;0 Beta Release &lpar;QCubed 3&period;0&period;0&rpar;</span></a></div>
```
It is the same problem with `_p` in non-human readable parts like href or src.
 - [X] The example `assets/php/examples/events_actions/javascript_alerts.php` has a link named `plugin that offers pre-built QDialogBox subclasses` with href like `http://localhost/example.com/vendor/qcubed/framework/plugins/DialogBoxes/example/more_dialog_boxes.php`. It resolves to the not found page in the default installation. If plugin is not installed, it should resolve to some `Install this plugin` page.
 - [X] Little 'bad practice' in example `/assets/php/examples/events_actions/event_delegation.php`?
```php
$this->dtgPersonsDelegated->AddAction($objOnEvent, new QJavaScriptAction("event.stopPropagation()"));
```
The example `assets/php/examples/events_actions/event_propagation.php` suggests use of the `QStopPropagationAction` wrapper:
```php
 $this->objPanel4->AddAction(new QClickEvent(), new QStopPropagationAction());
```
It's use should be consistent in all qcubed examples.
 - [X] The `assets/php/examples/events_actions/watcher.php` example suggests: `See the QJsTimer example page for help`, but provides no link to this example. Link should be provided.
 - [X] `assets/php/examples/datagrid/links_images.php` 
```
Error in PHP Script /example.com/vendor/qcubed/framework/assets/php/examples/datagrid/links_images.php
Incorrectly formatted DataGrid HTML in c1: <?= $_FORM->renderImage($_ITEM->Id) ?>
copy(/var/www/example.com/vendor/qcubed/framework/assets/php/examples/datagrid/cache/dc8e508515db386bdc31ee30214242cd.png): failed to open stream: Permission denied

Error Type: E_WARNING
```
The `assets/php/examples/datagrid/cache` should be writable. composer file should require that, and configuration checker should check it.
 - [X] `assets/php/examples/datagrid/filtering.php` has a link named `See the Introduction to Meta DataGrids for more information` with URL like `http://localhost/example.com/vendor/qcubed/framework/assets/php/examples/other/meta_datagrids.php`, which does not exists.
 - [X] `assets/php/examples/datagrid/ajax.php` the example is named `Enabling AJAX-based Sorting and Pagination`, but it has only one page, so user can not see pagination in action here. There are 12 items in a table. we should set ItemsCount to 5 to have at least 3 pages. Same issue for the `assets/php/examples/datagrid/extend.php` example. it is all about pagination, but has a datagrid with only one page.
 - [X] `assets/php/examples/master_detail/project_list.php` 
```
Error in PHP Script /example.com/vendor/qcubed/framework/assets/php/examples/master_detail/project_list.php
require_once(/var/www/example.com/project/includes/connector/ProjectDataGrid.class.php): failed to open stream: No such file or directory

Error Type: E_WARNING
```
 - [X] The example `assets/php/examples/other_controls/persist.php` named `Persistent Controls` does not work in a default installation. The `Reload the Page` button effectively resets both panels with saved and non-saved states.
 - [X] The example `assets/php/examples/image_label/intro.php` fails with message:
```
Fatal error: Call to undefined function imagepsloadfont() in /example.com/vendor/qcubed/framework/includes/base_controls/QImageLabelBase.class.php on line 179
```

My PHP version:
```
>$ php -v
PHP 5.5.9-1ubuntu4.11 (cli) (built: Jul  2 2015 15:23:08) 
Copyright (c) 1997-2014 The PHP Group
Zend Engine v2.5.0, Copyright (c) 1998-2014 Zend Technologies
    with Zend OPcache v7.0.3, Copyright (c) 1999-2014, by Zend Technologies
```

My box:
```
uname -a
Linux kremlrus 3.19.0-28-generic #30~14.04.1-Ubuntu SMP Tue Sep 1 09:32:55 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
```

There is a very-old thread about this problem on a forum: http://qcu.be/content/examples-error-imagepsloadfont-undefined.

Note that @Gianni has two points in his last comment.

I propose to change the default font used to something with more abroad support.

 - [X] ~~The `assets/php/examples/treenav/treenav.php` example does not loads anything. On directory expand it says `Loading...` and loads forever.~~ See https://github.com/qcubed/framework/issues/869#issuecomment-170064863.
 - [X] `assets/php/examples/other_controls/image.php`, `assets/php/examples/other_controls/file_asset.php` and `assets/php/examples/other_controls/image_file_asset.php`
```
Error in PHP Script /example.com/vendor/qcubed/framework/assets/php/examples/other_controls/image.php
imagejpeg(): Unable to open '/var/www/example.com/project/assets/images/cache/0badede937f701cc9cf35f644fc03b3a.jpg' for writing: No such file or directory
```
 - [X] `assets/php/examples/image_browser/test_image_browser.php`
```
Error in PHP Script /example.com/vendor/qcubed/framework/assets/php/examples/image_browser/test_image_browser.php
imagepng(): Unable to open '/var/www/example.com/vendor/qcubed/framework/assets/php/examples/image_browser/cache/dda460fc00c544f390acfe65fcf964a2.png' for writing: Permission denied
```
 - [X] `assets/php/examples/other_controls/timer_js.php` when loaded, it heats my CPU by 100% load with both, apache and browser processes one by one. If Stop is pressed, the heating stopped. May be it has too many items to update? There are 329 items for now.
 - [X] `assets/php/examples/other/datagrid_connectors.php` 
```
Error in PHP Script /example.com/vendor/qcubed/framework/assets/php/examples/other/datagrid_connectors.php
Class 'ProjectList' not found

Error Type: E_ERROR
```
 - ~~`assets/php/examples/basic_qform/xss.php`~~
```
Error in PHP Script /example.com/vendor/qcubed/framework/assets/php/examples/basic_qform/xss.php
require_once(/var/www/example.com/vendor/ezyang/htmlpurifier/library/HTMLPurifier.auto.php): failed to open stream: No such file or directory
```
~~htmlpurifier should be required by the composer file?~~
 - But the HTMLPurifier protection does not works. It works with default settings, but fails to protect if some tags are allowed: `Textbox protected with QCrossScripting::HTMLPurifier with a set of tags that's allowed (ex.<b>, <i>)` moved to #880
 - [X] `assets/php/examples/plugins/components.php` The function in this: `GetModelConnectorParamsfunction` should be separated with a space.

And the last, but not least: I have a general issue for plugin system #868

**Comment by olegabr** [comment]
@spekary, why the AddressConnector has no lblId_Create method anymore?
It is used in the `assets/php/examples/master_detail/project_list.php` example. I have a fix for this example, but I've just removed all occurenced of AddressConnector::lblId_Create in it. I'm not sure that it is the right fix. May be the bug is that the method lblId_Create is not generated? Is it by design or by accident?

**Comment by kukrik** [comment]
@spekary, here is one more mistake ...

I do not know if this is only a problem for me. But the example http://localhost/qcubed-shannon/vendor/qcubed/framework/assets/php/examples/communication/soap.php does not work correctly, the address is not correct, one has a double slash http:// next. I put the error message here:
```
Notice: Undefined index: HTTP_HOST in /Users/.../.../WEB/qcubed-shannon/vendor/qcubed/framework/assets/php/examples/communication/soap.php on line 48

Call Stack:
    0.0004     233936   1. {main}() /Users/.../.../WEB/qcubed-shannon/vendor/qcubed/framework/assets/php/examples/communication/soap.php:0


Fatal error: Uncaught SoapFault exception: [WSDL] SOAP-ERROR: Parsing WSDL: Couldn't load from 'http:///qcubed-shannon/vendor/qcubed/framework/assets/php/examples/communication/example_service.php?wsdl' : failed to load external entity "http:///qcubed-shannon/vendor/qcubed/framework/assets/php/examples/communication/example_service.php?wsdl"
 in /Users/.../.../WEB/qcubed-shannon/vendor/qcubed/framework/assets/php/examples/communication/soap.php:54
Stack trace:
#0 /Users/.../.../WEB/qcubed-shannon/vendor/qcubed/framework/assets/php/examples/communication/soap.php(54): SoapClient->SoapClient('http:///qcubed-...')
#1 {main}
  thrown in /.../.../WEB/qcubed-shannon/vendor/qcubed/framework/assets/php/examples/communication/soap.php on line 54

Process finished with exit code 255
```
Any thoughts?

**Comment by olegabr** [comment]
The AddressConnector::lblId_Create method is not generated because it is explicitly denied in a file `project/includes/configuration/codegen_options.json`:
```
    "Address": {
        "Id": {
            "FormGen": "none"
        }
}
```
I assume it was by design, so, my fix for the `assets/php/examples/master_detail/project_list.php` example is correct - "I've just removed all occurenced of AddressConnector::lblId_Create in it"

**Comment by olegabr** [comment]
Commit https://github.com/olegabr/framework/commit/421d04dcfbd3f6cd9307924258978baed24ad84b fixed a problem with assets/php/examples/other/datagrid_connectors.php example. It loads successfully now. But the column `PROJECTSTATUSTYPE` shows values like `<strong>1</strong>`

**Comment by olegabr** [comment]
Fix for `assets/php/examples/basic_qform/xss.php` issue: https://github.com/qcubed/app-starter/pull/11

**Comment by olegabr** [comment]
But the HTMLPurifier protection does not works. It works with default settings, but fails to protect if some tags are allowed: `Textbox protected with QCrossScripting::HTMLPurifier with a set of tags that's allowed (ex.<b>, <i>)`

**Comment by olegabr** [comment]
I've found the source of this issue: "The assets/php/examples/treenav/treenav.php example does not loads anything. On directory expand it says Loading... and loads forever."

There is an optimisation here:
https://github.com/qcubed/framework/blob/beta-3.0/includes/base_controls/QFormBase.class.php#L472
and it is buggy. I've removed it completely in my local code and the treenav example works as expected.

This optimisation is really important, but some listener pattern should be used to update only those controls that needs it. This should be merged with #631 I believe. The #631 should be implemented in a way that doesn't requires to iterate over all controls on the form. And Form should use the same technique to call ParsePostData only on those controls that really needs it.

I propose to postpone this optimization and the #631 issue to the v3.1 milestone. And for now I'm going to just comment the buggy optimization code out.

**Comment by spekary** [comment]
The QForm code is not buggy. The problem is that the TreeNav was trying to use a hack. It was manually issuing an expand event through a js call to PostAjax, without actually creating an expand event, and then using the fact that ParsePostData was getting called on all controls to look for the event and process the event. If you don't create an event on the php side, it will not normally get processed, even if you trigger the event in js code. The TreeNav in general does not use the best js code, but it works.

This is a general difficulty with the framework -- which is creating js widgets that communicate with the PHP code. The problem is that sometimes we are retrofitting js widgets that expect to work in a certain way, like JQ UI widgets, or Bootstrap widgets, and other times we are able to control both sides.

The fix is pretty easy and is posted in #874. It simply makes sure that a manual postAjax call will make sure that ParsePostData will get called on the PHP side for that particular control, which I think is reasonable. We should NOT be removing the optimization just because one widget doesn't work. We should in general fix the widget.

I would propose moving the QTreeNav widget to a plugin. There is no corresponding HTML or jQueryUI tree nav widget. There really is no reason to keep this in the core.

**Comment by olegabr** [comment]
@spekary, agreed with you completely. Merge the #874 please. I have no rights to do that.

**Comment by olegabr** [comment]
@spekary, I've investigated the problem with `other_controls/persist.php` example. It has following form data in AJAX call:
```
&c1=c1_1&c3=Change Me1&c2=c2_2&c4=Change Me2&Qform__FormId=PersistentExampleForm&Qform__FormControl=c5&Qform__FormEvent=QClickEvent#a1&Qform__FormParameter=&Qform__FormCallType=Ajax&Qform__FormUpdates=&Qform__FormCheckableControls=&Qform__FormState=56913216cb6ba_1
```

Current code in QFormBase::Run method does not understand that `c3=Change Me1` means that value of the `c3` control should be changed to the `Change Me1`. Same for listboxes (`c1=c1_1`).

It seems that we have two different interfaces to post control changes back to the server - the `Qform__FormUpdates` parameter and values from controls themselves, like the `c1=c1_1`. And these two mechanizms are not equally supported.

Do we need both of them, or one of them can be eliminated? If so, wich one should survive?

**Comment by spekary** [comment]
@olegabr Yes we need both of them. When a form's Post method is performed, the browser will automatically gather up all values of the html form objects and create variables for them. This process leaves out additional information that javascript controls might need to send to the server (for example, the value of the control's Parent control id if that was changed by a javascript widget). The formUpdates() function in qcubed.js is what is responsible for gathering up all the additional changes that need to be posted to the server. Those get sent through the Qform__FormUpdates hidden input control, and eventually com through as php properties through the __set mechanism, which happens on this line in QFormBase 
```
$objClass->objControlArray[$strControlId]->__set($strProperty, $strValue);
```
When an ajax method is performed, qcubed.js has to gather up the formUpdates() AND the post variables that the browser would normally gather in a form Post, so that our FormBase code is presented with the same information, and can deal with it generally in the same way.

I made an ajax optimization that only sent Post data to controls listed in the Post variables. I was trying to make the ajax process as fast as possible, since even small delays in ajax processing can cause a bad user experience for some controls. That optimization would break controls that were using the ParsePostData function to look at data that was not actually form post data directed to them. The TreeNav was the first control we found that did this, but fortunately, it was easy to fix.

You said "these two mechanizms are not equally supported.". They should be. They are supported in different ways, but they should both be supported, and we need to fix problems we find that don't support both.
